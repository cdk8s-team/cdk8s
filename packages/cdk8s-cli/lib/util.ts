import * as http from 'http';
import * as https from 'https';
import { spawn, SpawnOptions } from 'child_process';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

export async function shell(program: string, args: string[] = [], options: SpawnOptions = { }) {
  return new Promise((ok, ko) => {
    const child = spawn(program, args, { stdio: 'inherit', ...options });
    child.once('error', ko);
    child.once('exit', code => {
      if (code === 0) { return ok(); }
      else { return ko(new Error(`non-zero exit code ${code}`)); }
    });
  });
}

export async function withTempDir(dirname: string, closure: () => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'cdk8s.'));
  const workdir = path.join(parent, dirname);
  await fs.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fs.remove(parent);
  }
}

async function get(url: string, protocol: typeof http | typeof https = https): Promise<string> {
  return new Promise((ok, ko) => {
    const req = protocol.get(url, async res => {
      switch(res.statusCode) {
        case 200: {
          const data = new Array<Buffer>();
          res.on('data', chunk => data.push(chunk));
          res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
          res.once('error', ko);
          break;
        }

        case 301:
        case 302: {
          if (res.headers.location) {
            await ok(get(res.headers.location, protocol));
          }
          break;
        }

        default: {
          throw new Error(`${res.statusMessage}: ${url}`);
        }
      }
    });

    req.once('error', ko);
    req.end();
  });
}

export async function httpGet(url: string): Promise<string> {
  return get(url, http)
}

export async function httpsGet(url: string): Promise<string> {
  return get(url)
}
