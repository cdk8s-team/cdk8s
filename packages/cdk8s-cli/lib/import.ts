import * as https from 'https';
import * as fs from 'fs-extra';
import * as path from 'path';
import { promises } from 'fs';
import { emitConstructForApiObject, findApiObjectDefinitions, selectApiObjects } from '../lib/codegen-constructs';
import { CodeMaker } from 'codemaker';
import { JSONSchema4 } from 'json-schema';
import { TypeGenerator } from './codegen-types';
import { withTempDir, shell } from './util';
import { jsiiCompile } from './jsii';

export const DEFAULT_API_VERSION = '1.14.0';

export enum Language {
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  DOTNET = 'dotnet',
  JAVA = 'java',
}

export const LANGUAGES = [ Language.TYPESCRIPT, Language.PYTHON ];

export interface Options {
  /**
   * Output programming language.
   */
  readonly language: Language;

  /**
   * The API version to generate.
   */
  readonly apiVersion?: string;

  /**
   * FQNs of API object types to select instead of selecting the latest stable
   * version.
   * 
   * @default - selects the latest stable version from each API object
   */
  readonly include?: string[];

  /**
   * Do not import these types. Instead, represent them as "any".
   * 
   * @default - include all types that derive from the root types.
   */
  readonly exclude?: string[];
}

export async function generateAllApiObjects(outdir: string, options: Options) {
  outdir = path.resolve(outdir);

  if (options.language === Language.TYPESCRIPT) {
    await generateApiObjectsTypeScript(outdir, options);
    console.log(`${outdir}/k8s.ts`);
    return;
  }

  // this is not typescript, so we generate in a staging directory and harvest the code
  await withTempDir('k8s', async () => {
    await generateApiObjectsTypeScript('.', options);
    await jsiiCompile('.');

    const pacmak = require.resolve('jsii-pacmak/bin/jsii-pacmak');
    await shell(pacmak, [ '--target', options.language, '--code-only' ]);
    await harvestCode(options.language, outdir);
  });
}

async function harvestCode(language: Language, outdir: string) {
  switch (language) {
    case Language.TYPESCRIPT:
      throw new Error('no op for typescript');

    case Language.PYTHON:
      await harvestPython();
      break;

    default:
      throw new Error(`unsupported language ${language} (yet)`);
  }

  async function harvestPython() {
    const target = path.join(outdir, 'k8s');
    await fs.move('dist/python/src/k8s', target, { overwrite: true });
    console.error(target);
  }
}

async function generateApiObjectsTypeScript(outdir: string, options: Options) {
  const code = new CodeMaker();
  code.indentation = 2;

  const schema = await downloadSchema(options.apiVersion ?? DEFAULT_API_VERSION);
  const map = findApiObjectDefinitions(schema);

  const topLevelObjects = selectApiObjects(map, { include: options.include });

  code.openFile('k8s.ts');
  code.line(`// generated by cdk8s`);
  code.line(`import { ApiObject } from 'cdk8s';`);
  code.line(`import { Construct } from '@aws-cdk/core';`);
  code.line();

  const typeGenerator = new TypeGenerator(schema, { exclude: options.exclude });

  for (const o of topLevelObjects) {
    emitConstructForApiObject(code, typeGenerator, o);
  }

  typeGenerator.generate(code);

  code.closeFile('k8s.ts');

  await promises.mkdir(outdir, { recursive: true });
  await code.save(outdir);  
}

async function downloadSchema(apiVersion: string) {
  const output = await httpsGet(`https://kubernetesjsonschema.dev/v${apiVersion}/_definitions.json`);
  return JSON.parse(output) as JSONSchema4;
}

async function httpsGet(url: string): Promise<string> {
  return new Promise((ok, ko) => {
    const req = https.get(url, res => {
      if (res.statusCode !== 200) {
        throw new Error(`${res.statusMessage}: ${url}`);
      }
      const data = new Array<Buffer>();
      res.on('data', chunk => data.push(chunk));
      res.once('end', () => ok(Buffer.concat(data).toString('utf-8')));
      res.once('error', ko);
    });

    req.once('error', ko);
    req.end();
  });
}
