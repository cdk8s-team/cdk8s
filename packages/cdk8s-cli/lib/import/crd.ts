import { TypeGenerator } from './type-generator';
import { ImportBase } from './base';
import { CodeMaker } from 'codemaker';
import { httpsGet } from '../util';
import * as yaml from 'yaml';
import * as fs from 'fs-extra';
import * as path from 'path';

export interface CustomResourceApiObject {
  apiVersion?: string;
  kind?: string;
  metadata?: {
    name?: string;
  };
  spec?: {
    group: string;
    names: {
      kind: string;
    };
    versions?: Array<{ name: string; schema?: { openAPIV3Schema?: any } }>;
    validation?: { openAPIV3Schema?: any };
  };
}

export class CustomResourceDefinition {
  private readonly schema: any;
  private readonly group: string;
  private readonly version: string;
  readonly kind: string;
  private readonly fqn: string;

  constructor(manifest: CustomResourceApiObject) {
    assert(manifest.apiVersion === 'apiextensions.k8s.io/v1beta1', '"apiVersion" must be "apiextensions.k8s.io/v1beta1"');
    assert(manifest.kind === 'CustomResourceDefinition', '"kind" must be "CustomResourceDefinition"');

    const spec = manifest.spec;
    if (!spec) {
      throw new Error(`manifest does not have a "spec" attribute`)
    }
  
    const version = (spec.versions ?? [])[0];
    if (!version) {
      throw new Error(`unable to determine CRD version`);
    }
  
    if (!manifest.metadata?.name) {
      throw new Error(`"metadata.name" is required`);
    }
  
    const schema = version.schema?.openAPIV3Schema ?? spec.validation?.openAPIV3Schema;
  
    if (!schema) {
      throw new Error(`unable to determine schema for custom resource`);
    }

    this.schema = schema;
    this.group = spec.group;
    this.version = version.name;
    this.kind = spec.names.kind;
    this.fqn = this.kind;
  }

  public get moduleName() {
    return this.kind.toLocaleLowerCase();
  }

  public async generateTypeScript(code: CodeMaker) {
    const types = new TypeGenerator();
  
    types.emitConstruct({
      group: this.group,
      version: this.version,
      kind: this.kind,
      fqn: this.fqn,
      schema: this.schema
    });

    code.line(`// generated by cdk8s`);
    code.line(`import { ApiObject } from 'cdk8s';`);
    code.line(`import { Construct } from 'constructs';`);
    code.line();
    types.generate(code);
  }
}

export class ImportCustomResourceDefinition extends ImportBase {
  public static async match(source: string): Promise<undefined | CustomResourceApiObject[]> {
    let manifest;
    if (source.startsWith('https://')) {
      manifest = await httpsGet(source);
    } else if (path.extname(source) === '.yaml' || path.extname(source) === '.yml' || path.extname(source) === '.json') {
      if (!(await fs.pathExists(source))) {
        throw new Error(`can't find file ${source}`);
      }

      manifest = await fs.readFile(source, 'utf-8');
    }

    if (!manifest) {
      return undefined;
    }

    return yaml.parseAllDocuments(manifest)
               .map((doc: yaml.ast.Document) => doc.toJSON())
               .filter((doc) => (doc as CustomResourceApiObject).kind === 'CustomResourceDefinition');
  }

  private readonly customResourceDefinitions: CustomResourceDefinition[] = [];
  
  constructor(manifest: CustomResourceApiObject[]) {
    super();

    this.customResourceDefinitions = manifest?.map(obj => new CustomResourceDefinition(obj));
  }

  public get moduleNames() {
    return this.customResourceDefinitions.map(crd => crd.moduleName);
  }

  protected async generateTypeScript(code: CodeMaker, moduleName: string) {
    this.customResourceDefinitions.filter(crd => moduleName === crd.moduleName).map(crd => crd.generateTypeScript(code));
  }
}

function assert(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(`invalid CustomResourceDefinition manifest: ${message}`);
  }
}
