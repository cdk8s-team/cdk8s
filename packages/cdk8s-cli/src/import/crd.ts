import { ImportBase } from './base';
import { CodeMaker } from 'codemaker';
import { download } from '../util';
import * as yaml from 'yaml';
import { ImportSpec } from '../config';
import { TypeGenerator } from 'json2jsii';
import { generateConstruct } from './codegen';

const CRD_KIND = 'CustomResourceDefinition';

export interface ManifestObjectDefinition {
  apiVersion?: string;
  kind?: string;
  items?: ManifestObjectDefinition[]; // if `kind` is "List"
  metadata?: {
    name?: string;
  };
  spec?: {
    group: string;
    names: {
      kind: string;
      [key: string]: any;
    };
    versions?: Array<{
      name: string;
      schema?: { openAPIV3Schema?: any };
      [key: string]: any;
    }>;
    version?: string;
    validation?: { openAPIV3Schema?: any };
    [key: string]: any;
  };
}

// all these APIs are compatible from our perspective.
const SUPPORTED_API_VERSIONS = [
  'apiextensions.k8s.io/v1beta1',
  'apiextensions.k8s.io/v1',
];

export class CustomResourceDefinition {
  private readonly schema?: any;
  private readonly group: string;
  private readonly version: string;
  private readonly kind: string;
  private readonly fqn: string;

  constructor(manifest: ManifestObjectDefinition) {
    const apiVersion = manifest?.apiVersion ?? 'undefined';
    assert(SUPPORTED_API_VERSIONS.includes(apiVersion), `"apiVersion" is "${apiVersion}" but it should be one of: ${SUPPORTED_API_VERSIONS.map(x => `"${x}"`).join(', ')}`);
    assert(manifest.kind === CRD_KIND, `"kind" must be "${CRD_KIND}"`);

    const spec = manifest.spec;
    if (!spec) {
      throw new Error('manifest does not have a "spec" attribute')
    }

    const version = spec.version ?? (spec.versions ?? [])[0];
    if (!version) {
      throw new Error('unable to determine CRD version');
    }

    const schema = typeof version === 'string'
      ? spec.validation?.openAPIV3Schema
      : version?.schema?.openAPIV3Schema ?? spec.validation?.openAPIV3Schema;

    this.schema = schema;
    this.group = spec.group;
    this.version = typeof version === 'string' ? version : version.name;
    this.kind = spec.names.kind;
    this.fqn = this.kind;
  }

  public get moduleName() {
    return `${this.group}/${this.kind.toLocaleLowerCase()}`;
  }

  public async generateTypeScript(code: CodeMaker) {
    const types = new TypeGenerator();

    generateConstruct(types, {
      group: this.group,
      version: this.version,
      kind: this.kind,
      fqn: this.fqn,
      schema: this.schema,
    });

    code.line('// generated by cdk8s');
    code.line('import { ApiObject } from \'cdk8s\';');
    code.line('import { Construct } from \'constructs\';');
    code.line();
    code.line(types.render());
  }
}

export class ImportCustomResourceDefinition extends ImportBase {
  public static async match(importSpec: ImportSpec): Promise<undefined | ManifestObjectDefinition[]> {
    const { source } = importSpec;
    const manifest = await download(source);
    return yaml.parseAllDocuments(manifest).map((doc: yaml.Document) => doc.toJSON());
  }

  private readonly defs: CustomResourceDefinition[] = [];

  constructor(manifest: ManifestObjectDefinition[]) {
    super();

    const defs: Record<string, CustomResourceDefinition> = { };

    const extractCRDs = (objects: ManifestObjectDefinition[] = []) => {
      for (const obj of objects) {
        // filter empty docs in the manifest
        if (!obj) {
          continue;
        }

        // found a crd, yey!
        if (obj.kind === CRD_KIND) {
          const crd = new CustomResourceDefinition(obj);
          const key = crd.moduleName;

          if (key in defs) {
            throw new Error(`${key} already exists`);
          }

          defs[key] = crd;
          continue;
        }

        // recurse into lists
        if (obj.kind === 'List') {
          extractCRDs(obj.items);
          continue;
        }
      }
    };

    extractCRDs(manifest);

    this.defs = Object.values(defs);
  }

  public get moduleNames() {
    return this.defs.map(crd => crd.moduleName);
  }

  protected async generateTypeScript(code: CodeMaker, moduleName: string) {
    for (const crd of this.defs.filter(crd => moduleName === crd.moduleName)) {
      await crd.generateTypeScript(code);
    }
  }
}

function assert(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(`invalid CustomResourceDefinition manifest: ${message}`);
  }
}
