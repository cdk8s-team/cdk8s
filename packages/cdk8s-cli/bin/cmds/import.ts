import * as yargs from 'yargs';
import { generateAllApiObjects, DEFAULT_API_VERSION } from '../../lib/import';

const K8S_SPEC = 'k8s';
const DEFAULT_OUTDIR = 'imports';

class Command implements yargs.CommandModule {
  public readonly command = 'import SPEC';
  public readonly describe = 'Imports API objects to your app by generating constructs';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', {  desc: `The API spec to import (only "k8s" is currently supported)` })
    .demand('SPEC')
    .example(`cdk8s import k8s`, `imports Kubenretes API objects to imports/k8s.ts`)
    .example(`cdk8s import k8s@1.13.0`, `imports a specific version of the Kubenretes API`)

    .option('output', { type: 'string', desc: 'output directory', default: DEFAULT_OUTDIR, alias: 'o' })
    .option('include', { type: 'array', desc: 'types to select instead of the default which is latest stable version' })
    .option('exclude', { type: 'array', desc: 'do not import types that match these regular expressions. They will be represented as the "any" type' });

  public async handler(argv: any) {
    const spec = argv.spec;
    const [ src, specver ] = spec.split('@');

    const apiVersion = specver ?? DEFAULT_API_VERSION;

    if (src !== K8S_SPEC) {
      throw new Error(`currently we only support the source ${K8S_SPEC}`);
    }

    await generateAllApiObjects(argv.output, {
      apiVersion,
      include: argv.include,
      exclude: argv.exclude
    });

    console.error(`k8s@${apiVersion} imported to ${argv.output}/k8s.ts`);
  }
}


module.exports = new Command();
