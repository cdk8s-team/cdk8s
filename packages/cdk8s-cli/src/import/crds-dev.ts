import { ImportCustomResourceDefinition, ManifestObjectDefinition } from './crd';

/**
 *
 *              github:crossplane/crossplane@0.14.0
 *              |--^- |  ^         ^         ^ ^  ^
 *                 |     |         |         | |  |
 *  - provider ----+     |         |         | |  |
 *  - account -----------+         |         | |  |
 *  - repo ------------------------+         | |  |
 *  - major ---------------------------------+ |  |
 *  - minor -----------------------------------+  |
 *  - patch --------------------------------------+
 */

/**
 * Matches a https://doc.crds.dev repo
 *
 *  - ManifestObjectDefinition[] if found
 *  - undefined if not
 *
 * @param source
 */
export async function importCrdsDevRepoMatch(source: string): Promise<undefined | ManifestObjectDefinition[]> {
  const url = crdsDevUrl(source);
  if (url) {
    const crd = await ImportCustomResourceDefinition.match({
      source: url,
    });

    return crd;
  }

  return undefined;
}

export function crdsDevUrl(source: string): (undefined | string) {
  const match = /^github:([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)(?:\@([0-9]+)\.([0-9]+)(?:\.([0-9]+))?)?$/.exec(source);
  if (match) {
    const account = match[1];
    const repo = match[2];
    const major = match[3];

    //default to master if no version specified
    //TODO: get latest released version from available versions
    let url = `https://doc.crds.dev/raw/github.com/${account}/${repo}`;
    if (major) {
      const minor = match[4];
      const patch = match[5] ?? '0';

      url = `https://doc.crds.dev/raw/github.com/${account}/${repo}@v${major}.${minor}.${patch}`;
    }

    return url;
  }

  return undefined;
}
