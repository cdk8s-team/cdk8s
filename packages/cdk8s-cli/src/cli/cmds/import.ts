import * as yargs from 'yargs';
import { readConfigSync, ImportSpec } from '../../config';
import { importDispatch } from '../../import/dispatch';

const config = readConfigSync();

const DEFAULT_OUTDIR = 'imports';
const LANGUAGES = [ 'typescript', 'python', 'java' ];

class Command implements yargs.CommandModule {
  public readonly command = 'import [SPEC]';
  public readonly describe = 'Imports API objects to your app by generating constructs.';
  public readonly aliases = [ 'gen', 'import', 'generate' ];

  public readonly builder = (args: yargs.Argv) => args
    .positional('SPEC', { default: config.imports, desc: 'import spec with the syntax [NAME:=]SPEC where NAME is an optional module name and supported SPEC are: k8s, crd.yaml, https://domain/crd.yaml).', array: true })
    .example('cdk8s import k8s', 'Imports Kubernetes API objects to imports/k8s.ts')
    .example('cdk8s import k8s@1.13.0', 'Imports a specific version of the Kubernetes API')
    .example('cdk8s import jenkins.io_jenkins_crd.yaml', 'Imports constructs for the Jenkins custom resource definition from a file')
    .example('cdk8s import mattermost:=mattermost_crd.yaml', 'Imports constructs for the mattermost cluster custom resource definition using a custom module name')

    .option('output', { default: DEFAULT_OUTDIR, type: 'string', desc: 'Output directory', alias: 'o' })
    .option('include', { type: 'array', desc: 'Types to select instead of the default which is latest stable version (only for "k8s")' })
    .option('exclude', { type: 'array', desc: 'Do not import types that match these regular expressions. They will be represented as the "any" type (only for "k8s")' })
    .option('language', { default: config.language, demand: true, type: 'string', desc: 'Output programming language', alias: 'l', choices: LANGUAGES });

  public async handler(argv: any) {
    let specs = Array.isArray(argv.spec) ? argv.spec : [ argv.spec ];

    specs = specs.filter((spec: string) => spec != null).map((spec: string) => parseImports(spec));

    await importDispatch(specs, argv, {
      outdir: argv.output,
      targetLanguage: argv.language,
    });
  }
}

function parseImports(spec: string): ImportSpec {
  const splitImport = spec.split(':=');

  // crd.yaml
  // url.com/crd.yaml
  if (splitImport.length === 1) {
    return {
      source: spec,
    }
  }

  // crd=crd.yaml
  // crd=url.com/crd.yaml
  if (splitImport.length === 2) {
    return {
      moduleNamePrefix: splitImport[0],
      source: splitImport[1],
    }
  }

  throw new Error('Unable to parse import specification. Syntax is [NAME:=]SPEC');
}

module.exports = new Command();
