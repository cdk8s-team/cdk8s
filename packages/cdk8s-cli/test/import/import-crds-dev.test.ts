import * as fs from 'fs';
import * as path from 'path';
import { mocked } from 'ts-jest/utils';
import { ImportCustomResourceDefinition } from '../../src/import/crd';
import { importCrdsDevRepoMatch, crdsDevUrl } from '../../src/import/crds-dev';
import { download } from '../../src/util';
import { expectImportMatchSnapshot } from './util';

jest.mock('../../src/util', () => {
  const mod = jest.requireActual('../../src/util');
  return {
    ...mod,
    download: jest.fn(),
  };
});

jest.setTimeout(3 * 60_000);

const fullImportTests = [
  { provider: 'github', account: 'crossplane', repo: 'crossplane', version: '0.14.0' },
];

const crdsDevUrlTests = [
  { import: 'github:crossplane/crossplane', expected: 'https://doc.crds.dev/raw/github.com/crossplane/crossplane' },
  { import: 'github:crossplane/crossplane@0.14', expected: 'https://doc.crds.dev/raw/github.com/crossplane/crossplane@v0.14.0' },
  { import: 'github:crossplane/crossplane@0.14.0', expected: 'https://doc.crds.dev/raw/github.com/crossplane/crossplane@v0.14.0' },
  { import: 'unsupported:crossplane/crossplane@0.14.0', expected: undefined, reason: 'unsupported provider' },
  { import: 'github:account/repo@1', expected: undefined, reason: 'major minor required if version present' },
];

const readFixture = (fixture: string) => {
  const fixtures = path.join(__dirname, 'fixtures', 'crds-dev');
  const filePath = path.join(fixtures, fixture);

  return fs.readFileSync(filePath, 'utf-8');
};

describe('crds.dev import', () => {
  for ( const t of fullImportTests ) {
    const source = `${t.provider}:${t.account}/${t.repo}@${t.version}`;

    mocked(download).mockImplementation((_url: string): Promise<string> => {
      return new Promise((ok, _ko) => {
        const mockFile = `${t.provider}-${t.account}-${t.repo}-v${t.version}.yaml`;
        const crds = readFixture(mockFile);
        ok(crds.toString());
      });
    });

    test(source, async () => {
      const crdsDev = await importCrdsDevRepoMatch(source);
      expect(crdsDev).toBeDefined();
      if (crdsDev) {
        await expectImportMatchSnapshot( () => new ImportCustomResourceDefinition(crdsDev) );
      }
    });
  }
});

describe('crds.dev import errors', () => {
  for ( const t of crdsDevUrlTests.filter( o => o.expected === undefined )) {
    test(t.import, async () => {
      const crdsDev = await importCrdsDevRepoMatch(t.import);
      expect(crdsDev).toBeUndefined();
    });
  }
});

describe('crds.dev url map', () => {
  for ( const t of crdsDevUrlTests ) {
    test(t.import, () => {
      const url = crdsDevUrl(t.import);
      expect(url).toBe(t.expected);
    });
  }
});
