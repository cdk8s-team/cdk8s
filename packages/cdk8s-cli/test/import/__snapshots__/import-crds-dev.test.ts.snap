// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crds.dev import github:crossplane/crossplane@0.14.0 emit module per API group 1`] = `
Object {
  "author": Object {
    "name": "generated@generated.com",
    "roles": Array [
      "author",
    ],
  },
  "dependencies": Object {
    "cdk8s": "999.999.999",
    "constructs": "3.3.71",
  },
  "dependencyClosure": Object {
    "cdk8s": Object {
      "targets": Object {
        "dotnet": Object {
          "namespace": "Org.Cdk8s",
          "packageId": "Org.Cdk8s",
        },
        "go": Object {
          "moduleName": "github.com/awslabs/cdk8s-go",
        },
        "java": Object {
          "maven": Object {
            "artifactId": "cdk8s",
            "groupId": "org.cdk8s",
          },
          "package": "org.cdk8s",
        },
        "js": Object {
          "npm": "cdk8s",
        },
        "python": Object {
          "distName": "cdk8s",
          "module": "cdk8s",
        },
      },
    },
    "constructs": Object {
      "targets": Object {
        "dotnet": Object {
          "namespace": "Constructs",
          "packageId": "Constructs",
        },
        "go": Object {
          "moduleName": "github.com/aws/constructs-go",
        },
        "java": Object {
          "maven": Object {
            "artifactId": "constructs",
            "groupId": "software.constructs",
          },
          "package": "software.constructs",
        },
        "js": Object {
          "npm": "constructs",
        },
        "python": Object {
          "distName": "constructs",
          "module": "constructs",
        },
      },
    },
  },
  "description": "metapkgcrossplaneio",
  "fingerprint": "<fingerprint>",
  "homepage": "http://generated",
  "jsiiVersion": "1.29.0 (build 41df200)",
  "license": "Apache-2.0",
  "metadata": Object {
    "jsii": Object {
      "pacmak": Object {
        "hasDefaultInterfaces": true,
      },
    },
  },
  "name": "metapkgcrossplaneio",
  "repository": Object {
    "type": "git",
    "url": "http://generated",
  },
  "schema": "jsii/0.10.0",
  "targets": Object {
    "js": Object {
      "npm": "metapkgcrossplaneio",
    },
  },
  "types": Object {
    "metapkgcrossplaneio.Configuration": Object {
      "assembly": "metapkgcrossplaneio",
      "base": "cdk8s.ApiObject",
      "docs": Object {
        "custom": Object {
          "schema": "Configuration",
        },
        "summary": "A Configuration is the description of a Crossplane Configuration package.",
      },
      "fqn": "metapkgcrossplaneio.Configuration",
      "initializer": Object {
        "docs": Object {
          "summary": "Defines a \\"Configuration\\" API object.",
        },
        "locationInModule": Object {
          "filename": "meta.pkg.crossplane.io.ts",
          "line": 39,
        },
        "parameters": Array [
          Object {
            "docs": Object {
              "summary": "the scope in which to define this object.",
            },
            "name": "scope",
            "type": Object {
              "fqn": "constructs.Construct",
            },
          },
          Object {
            "docs": Object {
              "summary": "a scope-local name for the object.",
            },
            "name": "id",
            "type": Object {
              "primitive": "string",
            },
          },
          Object {
            "docs": Object {
              "summary": "initialization props.",
            },
            "name": "props",
            "type": Object {
              "fqn": "metapkgcrossplaneio.ConfigurationProps",
            },
          },
        ],
      },
      "kind": "class",
      "locationInModule": Object {
        "filename": "meta.pkg.crossplane.io.ts",
        "line": 10,
      },
      "methods": Array [
        Object {
          "docs": Object {
            "remarks": "This can be used to inline resource manifests inside other objects (e.g. as templates).",
            "summary": "Renders a Kubernetes manifest for \\"Configuration\\".",
          },
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 26,
          },
          "name": "manifest",
          "parameters": Array [
            Object {
              "docs": Object {
                "summary": "initialization props.",
              },
              "name": "props",
              "type": Object {
                "fqn": "metapkgcrossplaneio.ConfigurationProps",
              },
            },
          ],
          "returns": Object {
            "type": Object {
              "primitive": "any",
            },
          },
          "static": true,
        },
      ],
      "name": "Configuration",
      "properties": Array [
        Object {
          "const": true,
          "docs": Object {
            "summary": "Returns the apiVersion and kind for \\"Configuration\\".",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 14,
          },
          "name": "GVK",
          "static": true,
          "type": Object {
            "fqn": "cdk8s.GroupVersionKind",
          },
        },
      ],
    },
    "metapkgcrossplaneio.ConfigurationProps": Object {
      "assembly": "metapkgcrossplaneio",
      "datatype": true,
      "docs": Object {
        "custom": Object {
          "schema": "Configuration",
        },
        "summary": "A Configuration is the description of a Crossplane Configuration package.",
      },
      "fqn": "metapkgcrossplaneio.ConfigurationProps",
      "kind": "interface",
      "locationInModule": Object {
        "filename": "meta.pkg.crossplane.io.ts",
        "line": 49,
      },
      "name": "ConfigurationProps",
      "properties": Array [
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "Configuration#spec",
            },
            "summary": "ConfigurationSpec specifies the configuration of a Configuration.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 60,
          },
          "name": "spec",
          "type": Object {
            "fqn": "metapkgcrossplaneio.ConfigurationSpec",
          },
        },
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "Configuration#metadata",
            },
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 53,
          },
          "name": "metadata",
          "optional": true,
          "type": Object {
            "primitive": "any",
          },
        },
      ],
    },
    "metapkgcrossplaneio.ConfigurationSpec": Object {
      "assembly": "metapkgcrossplaneio",
      "datatype": true,
      "docs": Object {
        "custom": Object {
          "schema": "ConfigurationSpec",
        },
        "summary": "ConfigurationSpec specifies the configuration of a Configuration.",
      },
      "fqn": "metapkgcrossplaneio.ConfigurationSpec",
      "kind": "interface",
      "locationInModule": Object {
        "filename": "meta.pkg.crossplane.io.ts",
        "line": 69,
      },
      "name": "ConfigurationSpec",
      "properties": Array [
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpec#crossplane",
            },
            "summary": "Semantic version constraints of Crossplane that package is compatible with.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 75,
          },
          "name": "crossplane",
          "optional": true,
          "type": Object {
            "fqn": "metapkgcrossplaneio.ConfigurationSpecCrossplane",
          },
        },
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpec#dependsOn",
            },
            "summary": "Dependencies on other packages.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 82,
          },
          "name": "dependsOn",
          "optional": true,
          "type": Object {
            "collection": Object {
              "elementtype": Object {
                "fqn": "metapkgcrossplaneio.ConfigurationSpecDependsOn",
              },
              "kind": "array",
            },
          },
        },
      ],
    },
    "metapkgcrossplaneio.ConfigurationSpecCrossplane": Object {
      "assembly": "metapkgcrossplaneio",
      "datatype": true,
      "docs": Object {
        "custom": Object {
          "schema": "ConfigurationSpecCrossplane",
        },
        "summary": "Semantic version constraints of Crossplane that package is compatible with.",
      },
      "fqn": "metapkgcrossplaneio.ConfigurationSpecCrossplane",
      "kind": "interface",
      "locationInModule": Object {
        "filename": "meta.pkg.crossplane.io.ts",
        "line": 91,
      },
      "name": "ConfigurationSpecCrossplane",
      "properties": Array [
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpecCrossplane#version",
            },
            "summary": "Semantic version constraints of Crossplane that package is compatible with.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 97,
          },
          "name": "version",
          "type": Object {
            "primitive": "string",
          },
        },
      ],
    },
    "metapkgcrossplaneio.ConfigurationSpecDependsOn": Object {
      "assembly": "metapkgcrossplaneio",
      "datatype": true,
      "docs": Object {
        "custom": Object {
          "schema": "ConfigurationSpecDependsOn",
        },
        "remarks": "One of Provider or Configuration may be supplied.",
        "summary": "Dependency is a dependency on another package.",
      },
      "fqn": "metapkgcrossplaneio.ConfigurationSpecDependsOn",
      "kind": "interface",
      "locationInModule": Object {
        "filename": "meta.pkg.crossplane.io.ts",
        "line": 106,
      },
      "name": "ConfigurationSpecDependsOn",
      "properties": Array [
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpecDependsOn#version",
            },
            "summary": "Version is the semantic version constraints of the dependency image.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 126,
          },
          "name": "version",
          "type": Object {
            "primitive": "string",
          },
        },
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpecDependsOn#configuration",
            },
            "summary": "Configuration is the name of a Configuration package image.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 112,
          },
          "name": "configuration",
          "optional": true,
          "type": Object {
            "primitive": "string",
          },
        },
        Object {
          "abstract": true,
          "docs": Object {
            "custom": Object {
              "schema": "ConfigurationSpecDependsOn#provider",
            },
            "summary": "Provider is the name of a Provider package image.",
          },
          "immutable": true,
          "locationInModule": Object {
            "filename": "meta.pkg.crossplane.io.ts",
            "line": 119,
          },
          "name": "provider",
          "optional": true,
          "type": Object {
            "primitive": "string",
          },
        },
      ],
    },
  },
  "version": "0.0.0",
}
`;
